### PR TITLE
feat: add market version to allow disabling markets

### LINKED ISSUE #1550: refactor(perp)!: Identify market with multi-part key [pair + version]

### PR DESCRIPTION
**Pull request by jgimeno**
- Closes #1550

### CODE REVIEW

File: x/common/testutil/genesis/perp_genesis.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ func AddPerpV2Genesis(gen app.GenesisState) app.GenesisState {
 		asset.Registry.Pair(denoms.BTC, denoms.NUSD): {
 			Market: perpv2types.Market{
 				Pair:                            asset.NewPair(denoms.BTC, denoms.NUSD),
+				Version:                         1,
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply x/common/testutil/genesis/perp_genesis.go:23]: I prefer the one given that 0 is the typical default value in proto

**Comment by jgimeno** [comment]
Let's be careful with this one, so we don't forget any edge case.

**Review by Unique-Divine — CHANGES_REQUESTED** [review]
Review up to `module/abci.go`. Will continue going through the rest after I wake up in the morning

File: x/perp/v2/keeper/admin.go [hunk 2]
```diff
@@ -71,28 +73,32 @@ func (k admin) CreateMarket(
...
 	}
 
+	lastVersion := k.MarketLastVersion.GetOr(ctx, pair, types.MarketLastVersion{Version: 0})
+	lastVersion.Version += 1
+	market.Version = lastVersion.Version
+
 	// init amm
 	amm := types.AMM{
 		Pair:            pair,
+		Version:         1,
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply x/perp/v2/keeper/admin.go:101]: good catch!

[the remark this answers is not in the document]
Comment (jgimeno) [reply x/perp/v2/keeper/admin.go]: fixed and added test to cover it.

File: x/perp/v2/keeper/amm.go [hunk 3]
```diff
@@ -39,7 +42,7 @@ func (k Keeper) EditPriceMultiplier(
 
 	// Do the re-peg
 	amm.PriceMultiplier = newPriceMultiplier
-	k.AMMs.Insert(ctx, pair, amm)
+	k.AMMs.Insert(ctx, collections.Join(pair, amm.Version), amm)
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply x/perp/v2/keeper/amm.go:45]: I agree in part, in theory entities should not know about implementation details of the other layer. In this case the persistence (keeper in our case) but i can do a convenience method in the keeper.

File: x/perp/v2/keeper/hooks.go [hunk 4]
```diff
@@ -17,7 +17,8 @@ func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochN
 }
 
 func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, _ uint64) {
-	for _, market := range k.Markets.Iterate(ctx, collections.Range[asset.Pair]{}).Values() {
+	for _, market := range k.Markets.Iterate(ctx, collections.Range[collections.Pair[asset.Pair, uint64]]{}).Values() {
+		// TODO (reviewer): this needs to be fixed for only enabled markets
```
[the remark this answers is not in the document]
Comment (jgimeno) [reply x/perp/v2/keeper/hooks.go:21]: I think is not needed given we already know if it is disabled

**Review by Unique-Divine — APPROVED** [review]

**Review by Unique-Divine — APPROVED** [review]

**Pull request merged**